README.md
# PyTorch-YOLOv3
A minimal PyTorch implementation of YOLOv3, with support for training, inference and evaluation.

## 5g-dive implementation
##### Requirements
1. Ubuntu 16.06
2. Conda Environment

##### Clone and install requirements
    $ git clone
    $ cd xxx
    $ conda create -n pytorch-yolov3 python=3.6 -c conda-forge
    
## Installation
##### Clone and install requirements
    $ git clone https://github.com/eriklindernoren/PyTorch-YOLOv3
    $ cd PyTorch-YOLOv3/
    $ sudo pip3 install -r requirements.txt

##### Download pretrained weights
    $ cd weights/
    $ bash download_weights.sh

##### Download COCO
    $ cd data/
    $ bash get_coco_dataset.sh
    
## Test
Evaluates the model on COCO test.

    $ python3 test.py --weights_path weights/yolov3.weights

| Model                   | mAP (min. 50 IoU) |
| ----------------------- |:-----------------:|
| YOLOv3 608 (paper)      | 57.9              |
| YOLOv3 608 (this impl.) | 57.3              |
| YOLOv3 416 (paper)      | 55.3              |
| YOLOv3 416 (this impl.) | 55.5              |

## Inference
Uses pretrained weights to make predictions on images. Below table displays the inference times when using as inputs images scaled to 256x256. The ResNet backbone measurements are taken from the YOLOv3 paper. The Darknet-53 measurement marked shows the inference time of this implementation on my 1080ti card.

| Backbone                | GPU      | FPS      |
| ----------------------- |:--------:|:--------:|
| ResNet-101              | Titan X  | 53       |
| ResNet-152              | Titan X  | 37       |
| Darknet-53 (paper)      | Titan X  | 76       |
| Darknet-53 (this impl.) | 1080ti   | 74       |

    $ python3 detect.py --image_folder data/samples/

<p align="center"><img src="assets/giraffe.png" width="480"\></p>
<p align="center"><img src="assets/dog.png" width="480"\></p>
<p align="center"><img src="assets/traffic.png" width="480"\></p>
<p align="center"><img src="assets/messi.png" width="480"\></p>

## Train
```
$ train.py [-h] [--epochs EPOCHS] [--batch_size BATCH_SIZE]
                [--gradient_accumulations GRADIENT_ACCUMULATIONS]
                [--model_def MODEL_DEF] [--data_config DATA_CONFIG]
                [--pretrained_weights PRETRAINED_WEIGHTS] [--n_cpu N_CPU]
                [--img_size IMG_SIZE]
                [--checkpoint_interval CHECKPOINT_INTERVAL]
                [--evaluation_interval EVALUATION_INTERVAL]
                [--compute_map COMPUTE_MAP]
                [--multiscale_training MULTISCALE_TRAINING]
```

#### Example (COCO)
To train on COCO using a Darknet-53 backend pretrained on ImageNet run: 
```
$ python3 train.py --data_config config/coco.data  --pretrained_weights weights/darknet53.conv.74
```

#### Training log
```
---- [Epoch 7/100, Batch 7300/14658] ----
+------------+--------------+--------------+--------------+
| Metrics    | YOLO Layer 0 | YOLO Layer 1 | YOLO Layer 2 |
+------------+--------------+--------------+--------------+
| grid_size  | 16           | 32           | 64           |
| loss       | 1.554926     | 1.446884     | 1.427585     |
| x          | 0.028157     | 0.044483     | 0.051159     |
| y          | 0.040524     | 0.035687     | 0.046307     |
| w          | 0.078980     | 0.066310     | 0.027984     |
| h          | 0.133414     | 0.094540     | 0.037121     |
| conf       | 1.234448     | 1.165665     | 1.223495     |
| cls        | 0.039402     | 0.040198     | 0.041520     |
| cls_acc    | 44.44%       | 43.59%       | 32.50%       |
| recall50   | 0.361111     | 0.384615     | 0.300000     |
| recall75   | 0.222222     | 0.282051     | 0.300000     |
| precision  | 0.520000     | 0.300000     | 0.070175     |
| conf_obj   | 0.599058     | 0.622685     | 0.651472     |
| conf_noobj | 0.003778     | 0.004039     | 0.004044     |
+------------+--------------+--------------+--------------+
Total Loss 4.429395
---- ETA 0:35:48.821929
```

#### Tensorboard
Track training progress in Tensorboard:
* Initialize training
* Run the command below
* Go to http://localhost:6006/

```
$ tensorboard --logdir='logs' --port=6006
```

## Train on Custom Dataset

#### Custom model
Run the commands below to create a custom model definition, replacing `<num-classes>` with the number of classes in your dataset.

```
$ cd config/                                # Navigate to config dir
$ bash create_custom_model.sh <num-classes> # Will create custom model 'yolov3-custom.cfg'
```

#### Classes
Add class names to `data/custom/classes.names`. This file should have one row per class name.

#### Image Folder
Move the images of your dataset to `data/custom/images/`.

#### Annotation Folder
Move your annotations to `data/custom/labels/`. The dataloader expects that the annotation file corresponding to the image `data/custom/images/train.jpg` has the path `data/custom/labels/train.txt`. Each row in the annotation file should define one bounding box, using the syntax `label_idx x_center y_center width height`. The coordinates should be scaled `[0, 1]`, and the `label_idx` should be zero-indexed and correspond to the row number of the class name in `data/custom/classes.names`.

#### Define Train and Validation Sets
In `data/custom/train.txt` and `data/custom/valid.txt`, add paths to images that will be used as train and validation data respectively.

#### Train
To train on the custom dataset run:

```
$ python3 train.py --model_def config/yolov3-custom.cfg --data_config config/custom.data
```

Add `--pretrained_weights weights/darknet53.conv.74` to train using a backend pretrained on ImageNet.


## Credit

### YOLOv3: An Incremental Improvement
_Joseph Redmon, Ali Farhadi_ <br>

**Abstract** <br>
We present some updates to YOLO! We made a bunch
of little design changes to make it better. We also trained
this new network that’s pretty swell. It’s a little bigger than
last time but more accurate. It’s still fast though, don’t
worry. At 320 × 320 YOLOv3 runs in 22 ms at 28.2 mAP,
as accurate as SSD but three times faster. When we look
at the old .5 IOU mAP detection metric YOLOv3 is quite
good. It achieves 57.9 AP50 in 51 ms on a Titan X, compared
to 57.5 AP50 in 198 ms by RetinaNet, similar performance
but 3.8× faster. As always, all the code is online at
https://pjreddie.com/yolo/.

[[Paper]](https://pjreddie.com/media/files/papers/YOLOv3.pdf) [[Project Webpage]](https://pjreddie.com/darknet/yolo/) [[Authors' Implementation]](https://github.com/pjreddie/darknet)

```
@article{yolov3,
  title={YOLOv3: An Incremental Improvement},
  author={Redmon, Joseph and Farhadi, Ali},
  journal = {arXiv},
  year={2018}
}
```


CHANGELOG.md
## 0.1.0 (一月 24, 2020)
  - initialization: added bump-version.sh and modified .gitignore
  - Merge pull request #156 from EverWinter23/patch-2
  - datasets.py: Convert image to RGB channel
  - Update README.md
  - Merge pull request #152 from timforby/patch-1
  - Preserve filenames of images produced by detect.py
  - allow for unscaled/non-normalized labels
  - Update README.md
  - Update README.md
  - Clean up
  - Common download weights script
  - Clean up
  - Update README.md
  - Update README.md
  - Merge pull request #151 from eriklindernoren/custom-training
  - README
  - README
  - README
  - Add support for loading backend weights pretrained on COCO
  - Docs
  - Merge branch 'custom-training' of https://github.com/eriklindernoren/PyTorch-YOLOv2 into custom-training
  - Docs
  - Merge branch 'master' into custom-training
  - Changed parameter names
  - Update README.md
  - Merge pull request #150 from timforby/master
  - Doc
  - Clean up
  - Removed custom model def.
  - Merge
  - Model evaluation after each epoch. Training on custom dataset instructions
  - Update README.md
  - Fix example label
  - Instructions for training on custom dataset
  - Fix error with compute_map, targets were expecting batch dimension but collate_fn reduces that dimension. Removed that slicing index and now works.
  - Clean up
  - Typo fix
  - Fix resize
  - Evaluates model on validation set after each epoch
  - Clean up
  - Multiscale : New size every tenth batch
  - More efficient multiscale training. Pillow instead of lycon
  - Added license
  - Merge branch 'master' of https://github.com/eriklindernoren/PyTorch-YOLOv
  - Added method for rescaling bounding boxes to original image
  - Update README.md
  - Update README.md
  - Update README.md
  - Fix Darknet.save_weights
  - Old var names
  - README
  - Clean up
  - Fixed test.py
  - Fix test.py and remove bbox iou compute for numpy arrays
  - Fixes bug in models.create_modules (shortcut and route layers)
  - Added if __name__ == '__main__'
  - Handles batch with no targets
  - Removed dev print
  - Removed use of Pillow. Updated requirements
  - Update README.md
  - Update README.md
  - Update README.md
  - Update README.md
  - Significantly faster training
  - Set loss parameters to balance precision / recall during training
  - Revert to pad 127.5
  - Faster. Resolves #139.
  - Logs recall50 and recall75. More efficient Tensorboard logging.
  - Fix Tensorboard logging
  - Update README.md
  - Update README.md
  - Update README.md
  - Changed log format
  - terminaltables for nicer logging
  - Multi-scale training
  - Merge branch 'master' of https://github.com/eriklindernoren/PyTorch-YOLOv3
  - Multi-scale training
  - Update README.md
  - Adds option to compute and log mAP during training
  - Removed CE
  - Use BCE as class loss (as in paper)
  - Gradient accumulation before step + clean up
  - Classification logging
  - Log approximate time left for epoch
  - Log expected time left for epoch
  - Clean up
  - Log losses (in Tensorboard and stdout) at each YOLO layer
  - Tensorboard
  - Merge pull request #137 from timforby/master
  - fix incorrect class loss calculation. cross entropy being averaged twice due to PyTorch defaults
  - Update README.md
  - Added score threshold when computing predictions for recall / precision
  - Added precision to train log
  - Balancing between obj and noobj part of confidence loss. Computation of mAP by class-wise average of APs.
  - Balancing between obj and noobj part of confidence loss. Computation of mAP by class-wise average of APs.
  - Balancing between obj and noobj part of confidence loss. Computation of mAP by class-wise average of APs.
  - Merge pull request #48 from id9502/master
  - Now tiny yolo-v3 can be used as detecting network, which is lightweight and can run about 5x faster than full Yolo-v3
  - Confidence mask according to paper using ignore threshold
  - Confidence mask according to paper using ignore threshold
  - Fix training bug when annotations are missing. Changed AP to recall
  - Added argument use_cuda
  - Computation of mAP
  - Merge branch 'master' of https://github.com/eriklindernoren/PyTorch-YOLO-V3
  - Fix to utils.bbox_iou. Stabilizes training. Resolves #25 #23 #18 #13 #12
  - Update README.md
  - Merge pull request #15 from 0asa/patch-1
  - Image folder argument was not used
  - Log progress in train.py
  - Bug fix in YOLOLayer.forward
  - Bug fix in YOLOLayer.forward
  - Merge pull request #8 from catalinolaru1/master
  - Added save_weights
  - Update README.md
  - Update README.md
  - Merge branch 'master' of https://github.com/eriklindernoren/PyTorch-YOLO-V3
  - torch.no_grad() in detect.py and test.py
  - Update README.md
  - Reproduced author's performance on COCO test
  - Update README.md
  - Update README.md
  - Update README.md
  - Merge
  - Reproduced author's performance on COCO test
  - Update README.md
  - Update README.md
  - Training refinements. Mostly work in YOLOLayer.forward and utils.build_targets
  - Training support! : )
  - Deleted root datasets.py. Updated installation instructions.
  - Support for loading coco. Preparation for training support
  - Simplified non-maximum suppression and faster test script.
  - Merge branch 'master' of https://github.com/eriklindernoren/PyTorch-YOLO-V3
  - CUDA support fix for non_max_suppression. Resolves #1
  - Update README.md
  - Update README.md
  - Update README.md
  - Removed output folder and replaced it with assets folder, which contains samples for README
  - Clean up
  - Added sample to README
  - Changed default confidence threshold for inference script
  - Cleaned up non-maximum suppression. Added arguments
  - Clean up
  - Created YOLO layer. Prep for training support
  - Update README.md
  - Update README.md
  - Update README.md
  - Update README.md
  - Update README.md
  - Refined image outputs
  - Initial commit
  - Initial commit



test.py
from __future__ import division

from models import *
from utils.utils import *
from utils.datasets import *
from utils.parse_config import *

import os
import sys
import time
import datetime
import argparse
import tqdm

import torch
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision import transforms
from torch.autograd import Variable
import torch.optim as optim


def evaluate(model, path, iou_thres, conf_thres, nms_thres, img_size, batch_size):
    model.eval()

    # Get dataloader
    dataset = ListDataset(path, img_size=img_size, augment=False, multiscale=False)
    dataloader = torch.utils.data.DataLoader(
        dataset, batch_size=batch_size, shuffle=False, num_workers=1, collate_fn=dataset.collate_fn
    )

    Tensor = torch.cuda.FloatTensor if torch.cuda.is_available() else torch.FloatTensor

    labels = []
    sample_metrics = []  # List of tuples (TP, confs, pred)
    for batch_i, (_, imgs, targets) in enumerate(tqdm.tqdm(dataloader, desc="Detecting objects")):

        # Extract labels
        labels += targets[:, 1].tolist()
        # Rescale target
        targets[:, 2:] = xywh2xyxy(targets[:, 2:])
        targets[:, 2:] *= img_size

        imgs = Variable(imgs.type(Tensor), requires_grad=False)

        with torch.no_grad():
            outputs = model(imgs)
            outputs = non_max_suppression(outputs, conf_thres=conf_thres, nms_thres=nms_thres)

        sample_metrics += get_batch_statistics(outputs, targets, iou_threshold=iou_thres)

    # Concatenate sample statistics
    true_positives, pred_scores, pred_labels = [np.concatenate(x, 0) for x in list(zip(*sample_metrics))]
    precision, recall, AP, f1, ap_class = ap_per_class(true_positives, pred_scores, pred_labels, labels)

    return precision, recall, AP, f1, ap_class


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", type=int, default=8, help="size of each image batch")
    parser.add_argument("--model_def", type=str, default="config/yolov3.cfg", help="path to model definition file")
    parser.add_argument("--data_config", type=str, default="config/coco.data", help="path to data config file")
    parser.add_argument("--weights_path", type=str, default="weights/yolov3.weights", help="path to weights file")
    parser.add_argument("--class_path", type=str, default="data/coco.names", help="path to class label file")
    parser.add_argument("--iou_thres", type=float, default=0.5, help="iou threshold required to qualify as detected")
    parser.add_argument("--conf_thres", type=float, default=0.001, help="object confidence threshold")
    parser.add_argument("--nms_thres", type=float, default=0.5, help="iou thresshold for non-maximum suppression")
    parser.add_argument("--n_cpu", type=int, default=4, help="number of cpu threads to use during batch generation")
    parser.add_argument("--img_size", type=int, default=416, help="size of each image dimension")
    opt = parser.parse_args()
    print(opt)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    print(" >>>> device = ", device)

    data_config = parse_data_config(opt.data_config)
    valid_path = data_config["valid"]
    class_names = load_classes(data_config["names"])

    # Initiate model
    model = Darknet(opt.model_def).to(device)
    if opt.weights_path.endswith(".weights"):
        # Load darknet weights
        model.load_darknet_weights(opt.weights_path)
    else:
        # Load checkpoint weights
        model.load_state_dict(torch.load(opt.weights_path))

    print("Compute mAP...")

    precision, recall, AP, f1, ap_class = evaluate(
        model,
        path=valid_path,
        iou_thres=opt.iou_thres,
        conf_thres=opt.conf_thres,
        nms_thres=opt.nms_thres,
        img_size=opt.img_size,
        batch_size=8,
    )

    print("Average Precisions:")
    for i, c in enumerate(ap_class):
        print(f"+ Class '{c}' ({class_names[c]}) - AP: {AP[i]}")

    print(f"mAP: {AP.mean()}")


detect.py
from __future__ import division

from models import *
from utils.utils import *
from utils.datasets import *

import os
import sys
import time
import datetime
import argparse

from PIL import Image

import torch
from torch.utils.data import DataLoader
from torchvision import datasets
from torch.autograd import Variable

import matplotlib.pyplot as plt
import matplotlib.patches as patches
from matplotlib.ticker import NullLocator

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--image_folder", type=str, default="data/samples", help="path to dataset")
    # parser.add_argument("--model_def", type=str, default="config/yolov3.cfg", help="path to model definition file")
    parser.add_argument("--model_def", type=str, default="config/yolo-obj-v5.cfg", help="path to model definition file")
    # parser.add_argument("--weights_path", type=str, default="weights/yolov3.weights", help="path to weights file")
    parser.add_argument("--weights_path", type=str, default="weights/yolo-dset6-best.weights", help="path to weights file")
    # parser.add_argument("--class_path", type=str, default="data/coco.names", help="path to class label file")
    parser.add_argument("--class_path", type=str, default="data/obj.names", help="path to class label file")
    # parser.add_argument("--conf_thres", type=float, default=0.8, help="object confidence threshold")
    parser.add_argument("--conf_thres", type=float, default=0.1, help="object confidence threshold")
    # parser.add_argument("--nms_thres", type=float, default=0.4, help="iou thresshold for non-maximum suppression")
    parser.add_argument("--nms_thres", type=float, default=0.1, help="iou thresshold for non-maximum suppression")
    parser.add_argument("--batch_size", type=int, default=1, help="size of the batches")
    parser.add_argument("--n_cpu", type=int, default=0, help="number of cpu threads to use during batch generation")
    # parser.add_argument("--img_size", type=int, default=416, help="size of each image dimension")
    parser.add_argument("--img_size", type=int, default=832, help="size of each image dimension")
    parser.add_argument("--checkpoint_model", type=str, help="path to checkpoint model")
    opt = parser.parse_args()
    print(opt)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    os.makedirs("output", exist_ok=True)

    # Set up model
    model = Darknet(opt.model_def, img_size=opt.img_size).to(device)

    if opt.weights_path.endswith(".weights"):
        # Load darknet weights
        model.load_darknet_weights(opt.weights_path)
    else:
        # Load checkpoint weights
        model.load_state_dict(torch.load(opt.weights_path))

    model.eval()  # Set in evaluation mode

    dataloader = DataLoader(
        ImageFolder(opt.image_folder, img_size=opt.img_size),
        batch_size=opt.batch_size,
        shuffle=False,
        num_workers=opt.n_cpu,
    )

    classes = load_classes(opt.class_path)  # Extracts class labels from file

    Tensor = torch.cuda.FloatTensor if torch.cuda.is_available() else torch.FloatTensor

    imgs = []  # Stores image paths
    img_detections = []  # Stores detections for each image index

    print("\nPerforming object detection:")
    prev_time = time.time()
    for batch_i, (img_paths, input_imgs) in enumerate(dataloader):
        # Configure input
        input_imgs = Variable(input_imgs.type(Tensor))

        # Get detections
        with torch.no_grad():
            detections = model(input_imgs)
            detections = non_max_suppression(detections, opt.conf_thres, opt.nms_thres)

        # Log progress
        current_time = time.time()
        inference_time = datetime.timedelta(seconds=current_time - prev_time)
        prev_time = current_time
        print("\t+ Batch %d, Inference Time: %s" % (batch_i, inference_time))

        # Save image and detections
        imgs.extend(img_paths)
        img_detections.extend(detections)

    # Bounding-box colors
    cmap = plt.get_cmap("tab20b")
    colors = [cmap(i) for i in np.linspace(0, 1, 20)]

    print("\nSaving images:")
    # Iterate through images and save plot of detections
    for img_i, (path, detections) in enumerate(zip(imgs, img_detections)):

        print("(%d) Image: '%s'" % (img_i, path))

        # Create plot
        img = np.array(Image.open(path))
        plt.figure()
        fig, ax = plt.subplots(1)
        ax.imshow(img)

        # Draw bounding boxes and labels of detections
        if detections is not None:
            # Rescale boxes to original image
            detections = rescale_boxes(detections, opt.img_size, img.shape[:2])
            unique_labels = detections[:, -1].cpu().unique()
            n_cls_preds = len(unique_labels)
            bbox_colors = random.sample(colors, n_cls_preds)
            for x1, y1, x2, y2, conf, cls_conf, cls_pred in detections:

                print("\t+ Label: %s, Conf: %.5f" % (classes[int(cls_pred)], cls_conf.item()))

                box_w = x2 - x1
                box_h = y2 - y1

                color = bbox_colors[int(np.where(unique_labels == int(cls_pred))[0])]
                # Create a Rectangle patch
                bbox = patches.Rectangle((x1, y1), box_w, box_h, linewidth=2, edgecolor=color, facecolor="none")
                # Add the bbox to the plot
                ax.add_patch(bbox)
                # Add label
                plt.text(
                    x1,
                    y1,
                    s=classes[int(cls_pred)],
                    color="white",
                    verticalalignment="top",
                    bbox={"color": color, "pad": 0},
                )

        # Save generated image with detections
        plt.axis("off")
        plt.gca().xaxis.set_major_locator(NullLocator())
        plt.gca().yaxis.set_major_locator(NullLocator())
        filename = path.split("/")[-1].split(".")[0]
        plt.savefig(f"output/{filename}.png", bbox_inches="tight", pad_inches=0.0)
        plt.close()


others/polygon.py
import fiona # pip install Fiona
from shapely.geometry import mapping, Polygon
# from shapely.geometry import Polygon # https://gis.stackexchange.com/questions/90055/finding-if-two-polygons-intersect-in-python
from shapely.geometry import shape # https://gis.stackexchange.com/questions/251812/returning-percentage-of-area-of-polygon-intersecting-another-polygon-using-shape

from PIL import Image, ImageDraw
im = Image.open("2020-01-24,15:42:26.tiff.jpg")
d = ImageDraw.Draw(im)

# line satu
l1_top_left = (120, 200)
l1_bottom_left = (845, 1074)
l1_top_right = (140, 188)
l1_bottom_right = (1260, 1076)
line1 = Polygon([l1_bottom_left, l1_bottom_right, l1_top_right, l1_top_left]) # line1

line_color = (0, 0, 255)

d.line([l1_top_left, l1_bottom_left, l1_bottom_right, l1_top_right, l1_top_left], fill=line_color, width=2)

# line kedua
l2_top_left = (140, 188)
l2_bottom_left = (1260, 1076)
l2_top_right = (162, 191)
l2_bottom_right = (1668, 1052)
line2 = Polygon([l2_bottom_left, l2_bottom_right, l2_top_right, l2_top_left]) # line2

line_color = (0, 255, 255)

d.line([l2_top_left, l2_bottom_left, l2_bottom_right, l2_top_right, l2_top_left], fill=line_color, width=2)

# line ketiga`
l3_top_left = (162, 191)
l3_bottom_left = (1668, 1052)
l3_top_right = (193, 195)
l3_bottom_right = (1914, 970)
line3 = Polygon([l3_bottom_left, l3_bottom_right, l3_top_right, l3_top_left]) # line3

line_color = (250, 200, 155)

d.line([l3_top_left, l3_bottom_left, l3_bottom_right, l3_top_right, l3_top_left], fill=line_color, width=2)

# Dumy mobil
c1_top_left = (450, 460)
c1_bottom_left = (460, 535)
c1_top_right = (535, 460)
c1_bottom_right = (535, 540)
car = Polygon([c1_bottom_left, c1_bottom_right, c1_top_right, c1_top_left]) # car

line_color = (100, 155, 255)

d.line([c1_top_left, c1_bottom_left, c1_bottom_right, c1_top_right, c1_top_left], fill=line_color, width=2)

# try calculate the intersection
def percentage_intersection(car, lines):
    for i in range(len(lines)):
        if car.intersects(lines[i]):
            intersect_percent = (car.intersection(lines[i]).area/car.area)*100
            # intersect = lines[i].intersection(car)
            # print(intersect)
            # print(intersect.area)
            print(" >>> car intersects Line ", (i+1), " -->> percent = ", intersect_percent)
        else:
            print(" >>> car NOT intersects Line ", (i+1))

percentage_intersection(car, [line1, line2, line3])

im.save("drawn_grid.png")


detect_video.py
from __future__ import division

from models import *
from utils.utils import *
from utils.datasets import *

import os
import sys
import time
import datetime
import argparse

from PIL import Image

import torch
from torch.utils.data import DataLoader
from torchvision import datasets
from torch.autograd import Variable

import matplotlib.pyplot as plt
import matplotlib.patches as patches
from matplotlib.ticker import NullLocator
import cv2 as cv

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--image_folder", type=str, default="data/samples", help="path to dataset")
    parser.add_argument("--model_def", type=str, default="config/yolov3.cfg", help="path to model definition file")
    parser.add_argument("--weights_path", type=str, default="weights/yolov3.weights", help="path to weights file")
    parser.add_argument("--class_path", type=str, default="data/coco.names", help="path to class label file")
    parser.add_argument("--conf_thres", type=float, default=0.8, help="object confidence threshold")
    parser.add_argument("--nms_thres", type=float, default=0.4, help="iou thresshold for non-maximum suppression")
    parser.add_argument("--batch_size", type=int, default=1, help="size of the batches")
    parser.add_argument("--n_cpu", type=int, default=0, help="number of cpu threads to use during batch generation")
    parser.add_argument("--img_size", type=int, default=416, help="size of each image dimension")
    parser.add_argument("--checkpoint_model", type=str, help="path to checkpoint model")
    opt = parser.parse_args()
    print(opt)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    os.makedirs("output", exist_ok=True)

    # Set up model
    model = Darknet(opt.model_def, img_size=opt.img_size).to(device)

    if opt.weights_path.endswith(".weights"):
        # Load darknet weights
        model.load_darknet_weights(opt.weights_path)
    else:
        # Load checkpoint weights
        model.load_state_dict(torch.load(opt.weights_path))

    model.eval()  # Set in evaluation mode

    dataloader = DataLoader(
        ImageFolder(opt.image_folder, img_size=opt.img_size),
        batch_size=opt.batch_size,
        shuffle=False,
        num_workers=opt.n_cpu,
    )

    classes = load_classes(opt.class_path)  # Extracts class labels from file

    Tensor = torch.cuda.FloatTensor if torch.cuda.is_available() else torch.FloatTensor

    imgs = []  # Stores image paths
    img_detections = []  # Stores detections for each image index

    # Ardi: Use video instead
    print("\nReading video:")
    cap = cv.VideoCapture("data/5g-dive/customTest_MIRC-Roadside-5s.mp4")
    cv.namedWindow("Image", cv.WND_PROP_FULLSCREEN)
    cv.resizeWindow("Image", 1366, 768) # Enter your size
    while (cap.isOpened()):
        # ret = a boolean return value from getting the frame, frame = the current frame being projected in the video
        try:
            ret, frame = cap.read()

            cv.imshow("Image", frame)
        except:
            print("No more frame to show.")
            break

        if cv.waitKey(10) & 0xFF == ord('q'):
            break
    # The following frees up resources and closes all windows
    cap.release()
    cv.destroyAllWindows()



    # print("\nPerforming object detection:")
    # prev_time = time.time()
    # for batch_i, (img_paths, input_imgs) in enumerate(dataloader):
    #
    #     # print(" >>> img_paths = ", img_paths)
    #     # print(" >>> input_imgs = ", input_imgs)
    #
    #     # Configure input
    #     input_imgs = Variable(input_imgs.type(Tensor))
    #
    #     # Get detections
    #     with torch.no_grad():
    #         detections = model(input_imgs)
    #         detections = non_max_suppression(detections, opt.conf_thres, opt.nms_thres)
    #
    #     # Log progress
    #     current_time = time.time()
    #     inference_time = datetime.timedelta(seconds=current_time - prev_time)
    #     prev_time = current_time
    #     print("\t+ Batch %d, Inference Time: %s" % (batch_i, inference_time))
    #
    #     # Save image and detections
    #     imgs.extend(img_paths)
    #     img_detections.extend(detections)
    #
    #
    # # Bounding-box colors
    # cmap = plt.get_cmap("tab20b")
    # colors = [cmap(i) for i in np.linspace(0, 1, 20)]
    #
    # print("\nSaving images:")
    # # Iterate through images and save plot of detections
    # for img_i, (path, detections) in enumerate(zip(imgs, img_detections)):
    #
    #     print("(%d) Image: '%s'" % (img_i, path))
    #
    #     # Create plot
    #     img = np.array(Image.open(path))
    #     plt.figure()
    #     fig, ax = plt.subplots(1)
    #     ax.imshow(img)
    #
    #     # Draw bounding boxes and labels of detections
    #     if detections is not None:
    #         # Rescale boxes to original image
    #         detections = rescale_boxes(detections, opt.img_size, img.shape[:2])
    #         unique_labels = detections[:, -1].cpu().unique()
    #
    #         print(" >> total detections = ", len(detections))
    #         print(" >> total unique_labels = ", len(unique_labels))
    #
    #         n_cls_preds = len(unique_labels)
    #         bbox_colors = random.sample(colors, n_cls_preds)
    #         for x1, y1, x2, y2, conf, cls_conf, cls_pred in detections:
    #
    #             print("\t+ Label: %s, Conf: %.5f" % (classes[int(cls_pred)], cls_conf.item()))
    #
    #             box_w = x2 - x1
    #             box_h = y2 - y1
    #
    #             color = bbox_colors[int(np.where(unique_labels == int(cls_pred))[0])]
    #             # Create a Rectangle patch
    #             bbox = patches.Rectangle((x1, y1), box_w, box_h, linewidth=2, edgecolor=color, facecolor="none")
    #             # Add the bbox to the plot
    #             ax.add_patch(bbox)
    #             # Add label
    #             plt.text(
    #                 x1,
    #                 y1,
    #                 s=classes[int(cls_pred)],
    #                 color="white",
    #                 verticalalignment="top",
    #                 bbox={"color": color, "pad": 0},
    #             )
    #
    #     # Save generated image with detections
    #     plt.axis("off")
    #     plt.gca().xaxis.set_major_locator(NullLocator())
    #     plt.gca().yaxis.set_major_locator(NullLocator())
    #     filename = path.split("/")[-1].split(".")[0]
    #     plt.savefig(f"output/{filename}.png", bbox_inches="tight", pad_inches=0.0)
    #     plt.close()
